### ISSUE TITLE
FOOPFF: ActiveCode gives an error for 2.17.1-2

### ISSUE THREAD

**Issue by georgianahaldeman** [RunestoneInteractive/rs#346]
**What Course are you in**
[Foundations of Python Programming: Functions First](https://runestone.academy/ns/books/published/colgateuniversity_COSC101C_f23/fopp-ff.html)

**What Page were you on**
https://runestone.academy/ns/books/published/colgateuniversity_COSC101C_f23/simple-python-data_chapter-assessment.html
2.17.1-2

**What is your username**
ghaldeman

**Describe the bug**
Myself (instructor) and my students are getting errors when trying to run code in the window using ActiveCode for 2.17.1-2. The code looks good, I even copied and pasted code that ran in 2.17.3, but I still get the error.

**Traceback**
Error
An error occurred after the end of your code.
One possible reason is that you have an unclosed parenthesis or string.
Another possibility is that there is an error in the hidden test code.
Yet another is that there is an internal error.  The internal error message is: undefined

**Comment by bnmnetp** [comment]
Tagging @cfcurtis

**Comment by bnmnetp** [comment]
@cfcurtis It looks like there are indentation errors in the unittest code for this exercise.

**Comment by cfcurtis** [comment]
Thanks for the tag, we've had a few similar errors that snuck in thanks to some XML autoformatting that wasn't code aware.

**Comment by cfcurtis** [comment]
I fixed the indentation and text wrapping issues, but I'm still getting the error that the function `square` is not defined. @bnmnetp, how do I make sure that code persists in a page? I tried moving the function definition before the exercises but it didn't help. The code in question is here: https://github.com/MRU-COMP1701-dev/fopp-functions-first/blob/master/pretext/SimplePythonData/week1a2.ptx

Thanks! I'm guessing there are more of these somewhere that slipped through the conversion process.

**Comment by oscarlevin** [comment]
There were two changes I made to fix the issue.  I used the include="ac2_4_1" attribute on the program, and added a label to the program (which I set to the same string as the xmlid)

**Comment by cfcurtis** [comment]
@oscarlevin thank you! I was able to make that change on the "Function Calls" section, but the same does not work for the "Chapter Assessment" section. For whatever reason when I add an `include` inside an `exercise` I get a 404 not found (it seems to be looking for a file with the specified label name), but when I comment out the `exercise` wrapper it works as expected.

**Comment by bnmnetp** [comment]
In looking at oscar's code.  The program containing the square function is not inside an exercise.  But contains label="ac2_4_1". But the program that does use that code is inside an exercise.

**Comment by cfcurtis** [comment]
Could you point me towards that code? I was looking at [this file](https://github.com/oscarlevin/foppcs/blob/main/pretext/SimplePythonData/FunctionCalls.ptx), and none of the program samples with includes are wrapped in exercise tags. [This](https://github.com/oscarlevin/foppcs/blob/main/pretext/SimplePythonData/week1a2.ptx) is the one where adding the include doesn't work unless I comment out `<exercise>`.

**Comment by bnmnetp** [comment]
Yes, the functionCalls.ptx file is in the body of the book, whereas the week1a2.ptx file is meant to be more of an Exercises page.

@oscarlevin , @rbeezer can you think of a reason why an `@include` would not work if the `program` was in an `exercise`?  I have not looked at the html to see if the `data-include` is missing in that case.

**Comment by rbeezer** [comment]
I can't think of why that location should make a difference.  Looking at `week1a2.ptx`.  Is the `program` getting a `label`  attribute in your testing?

And/or the `xml:id` value `assess_ps_01_02_editor` looks very suspicious and may clash with an HTML `id` attribute that is being generated automatically with a `_editor` appended?  You shouldn't need to create derived values like this.

**Comment by cfcurtis** [comment]
@rbeezer Thanks for the reply. Yes, I'm giving `program` a label - it works when I comment out the `exercise` tags (as is current [here](https://github.com/MRU-COMP1701-dev/fopp-functions-first/blob/master/pretext/SimplePythonData/week1a2.ptx)), but when I uncomment them it doesn't work.

The `_editor` id seemed odd to me as well so I renamed it `square_def` with no change. According to my commit history the xml id was there in the original pretext conversion, so it does seem to be auto generated.

**Comment by bnmnetp** [comment]
Closing this one out. The original bug — indentation broken by XML autoformatting in the unit-test code — was fixed by @cfcurtis at the time. What the thread turned into afterwards is a PreTeXt authoring question: why an `@include` on a `program` doesn't resolve when the program sits inside an `<exercise>`. That belongs with PreTeXt and the book's own repository rather than here, and the course it was reported against (Fall 2023) is long finished.

@cfcurtis, if that `include`-inside-`exercise` behavior is still biting you, the PreTeXt repo is the better venue and @rbeezer was already partway into diagnosing it here.
